### PR TITLE
drivers:adc:ad4080: Fix addrs and improve driver

### PR DESCRIPTION
**Pull request by ribdp**
## Pull Request Description

1. Fix addresses for GPIO config registers
2. Add macros for GPIO enabling
3. Add enum for GPIO function selection

## PR Type
- [x] Bug fix (change that fixes an issue)
- [ ] New feature (change that adds new functionality)
- [ ] Breaking change (has dependencies in other repos or will cause CI to fail)

## PR Checklist
- [x] I have followed the [Coding style guidelines](http://analogdevicesinc.github.io/no-OS/drivers_guide.html#coding-style)
- [x] I have performed a self-review of the changes
- [x] I have commented my code, at least hard-to-understand parts
- [ ] I have build all projects affected by the changes in this PR
- [x] I have tested in hardware affected projects, at the relevant boards
- [x] I have signed off all commits from this PR
- [ ] I have updated the documentation (wiki pages, ReadMe etc), if applies

### CODE REVIEW

**Review by amiclaus — DISMISSED** [review]

**Review by amiclaus — COMMENTED** [review]
I suppose these are changes related to the new silicon version. lgtm.

**Comment by ribdp** [comment]
Hi, is this good to be merged?

File: drivers/adc/ad4080/ad4080.h [hunk 1]
```diff
@@ -216,6 +220,21 @@ enum ad4080_intf_ldo_pd {
...
+	AD4080_GPIO_FILTER_RESULT_READY,
+	AD4080_GPIO_HT_DETECT,
+	AD4080_GPIO_LT_DETECT,
+	AD4080_GPIO_STATUS_ALERT,
+	AD4080_GPIO_GPO_DATA,
+	AD4080_GPIO_FILTER_SYNC_INPUT,
+	AD4080_GPIO_EXT_EVENT_TRIGGER_FIFO,
+	AD4080_GPIO_CNV_INHIBIT_INPUT
+};
+
```
Comment (rbolboac) [inline drivers/adc/ad4080/ad4080.h:237]: I see these new definitions here but I do not see how these can be used... don't they need to be used in a driver API?

Comment (ribdp) [reply drivers/adc/ad4080/ad4080.h:237]: That makes sense. I've been using the ad4080_read/write functions in my application along with these definitions. I'll make a generic function out of that for GPIO configurations, and make it part of this API.

File: drivers/adc/ad4080/ad4080.h [hunk 2]
```diff
@@ -122,6 +122,10 @@
 #define AD4080_ANA_DIG_LDO_PD_MSK		NO_OS_BIT(1)
 #define AD4080_INTF_LDO_PD_MSK			NO_OS_BIT(0)
 
+/** AD4080_GPIO_CONFIG Bit Defintion  */
+#define AD4080_GPIO_EN_MSK(x)			NO_OS_BIT(x)
+#define AD4080_GPIO_SEL_MSK(x)			(NO_OS_GENMASK(3, 0) << (4 * (x%2)))
```
Comment (CiprianRegus) [inline drivers/adc/ad4080/ad4080.h:127]: `(x)` instead of `x`

**Comment by ribdp** [comment]
Changelog:

1. Fix addresses for GPIO config registers
2. Add masks for GPIO configuration bitfields
3. Add enums for GPIOs, output enable states and function selection
4. Add methods for accessing and configuring fifo watermark
5. Add GPIO configuration APIs
6. Use fifo mode init param in the driver init
7. Add gpio output enable and func select as init params
7. Perform GPIO initialization before data intf initialization

File: drivers/adc/ad4080/ad4080.h (hunk 1 above)
Comment (ribdp) [reply drivers/adc/ad4080/ad4080.h]: I've updated the driver with the API

File: drivers/adc/ad4080/ad4080.c [hunk 3]
```diff
@@ -879,6 +879,190 @@ int ad4080_get_fifo_mode(struct ad4080_dev *dev,
...
+{
+	int ret;
+
+	if (!dev || fifo_watermark > AD4080_FIFO_SIZE)
+		return -EINVAL;
+
+	ret = ad4080_write(dev,
+			   AD4080_REG_FIFO_WATERMARK,
+			   no_os_field_get(BYTE_ADDR_L, fifo_watermark));
+	if (ret) {
```
Comment (amiclaus) [inline drivers/adc/ad4080/ad4080.c:899]: the convention is the same as for linux drivers. If only one statement within a conditional expression, then no brackets needed.

File: drivers/adc/ad4080/ad4080.c [hunk 4]
```diff
@@ -879,6 +879,190 @@ int ad4080_get_fifo_mode(struct ad4080_dev *dev,
...
+			      uint16_t *fifo_watermark)
+{
+	int ret;
+	uint8_t val;
+
+	if (!dev || !fifo_watermark)
+		return -EINVAL;
+
+	ret = ad4080_read(dev, AD4080_REG_FIFO_WATERMARK, &val);
+	if (ret) {
```
Comment (amiclaus) [inline drivers/adc/ad4080/ad4080.c:924]: ditto.

File: drivers/adc/ad4080/ad4080.c (hunk 3 above)
Comment (ribdp) [reply drivers/adc/ad4080/ad4080.c]: Fixed

File: drivers/adc/ad4080/ad4080.c (hunk 4 above)
Comment (ribdp) [reply drivers/adc/ad4080/ad4080.c]: Fixed

**Comment by ribdp** [comment]
Changelog - 
1. Removed extraneous brackets

**Comment by ribdp** [comment]
Changelog - 
1. Add missing return value check

**Review by amiclaus — DISMISSED** [review]

**Comment by ribdp** [comment]
Hi, are there any more comments/suggestions on this PR?

File: drivers/adc/ad4080/ad4080.c [hunk 5]
```diff
@@ -1001,12 +1184,27 @@ int ad4080_init(struct ad4080_dev **device,
...
 	ret = ad4080_configuration_intf_init(dev, init_param);
 	if (ret)
 		goto error_spi;
 
+	/* GPIO Initialization */
+	for (gpio = AD4080_GPIO_0; gpio < NUM_AD4080_GPIO; gpio++) {
+		/* Apply GPIO Output Enable state */
+		ret = ad4080_set_gpio_output_enable(dev, gpio, init_param.gpio_op_enable[gpio]);
+		if (ret)
+			return ret;
```
Comment (rbolboac) [inline drivers/adc/ad4080/ad4080.c:1200]: goto error_spi instead of returning

File: drivers/adc/ad4080/ad4080.c [hunk 6]
```diff
@@ -1001,12 +1184,27 @@ int ad4080_init(struct ad4080_dev **device,
...
+
+		/* Apply GPIO Output Function selection */
+		ret = ad4080_set_gpio_output_func(dev, gpio, init_param.gpio_op_func_sel[gpio]);
+		if (ret)
+			return ret;
```
Comment (rbolboac) [inline drivers/adc/ad4080/ad4080.c:1205]: goto error_spi instead of returning

File: drivers/adc/ad4080/ad4080.c [hunk 7]
```diff
@@ -1001,12 +1184,27 @@ int ad4080_init(struct ad4080_dev **device,
 
 	/* Set Operation mode */
 	ret = ad4080_set_op_mode(dev, init_param.op_mode);
+	if (ret)
+		return ret;
```
Comment (rbolboac) [inline drivers/adc/ad4080/ad4080.c:1188]: goto error_spi instead of returning

**Review by rbolboac — COMMENTED** [review]
some remarks from my side, also, you have build warnings:
```
CC ad4080.c
adc/ad4080/ad4080.c: In function 'ad4080_set_gpio_output_enable':
adc/ad4080/ad4080.c:950:6: warning: variable 'ret' set but not used [-Wunused-but-set-variable]
  950 |  int ret;
      |      ^~~
adc/ad4080/ad4080.c: In function 'ad4080_set_gpio_output_func':
adc/ad4080/ad4080.c:975:6: warning: variable 'ret' set but not used [-Wunused-but-set-variable]
  975 |  int ret;
      |      ^~~
```
please fix them accordingly.
Also, you only have one commit which has a list of things done.. in general a commit should be related to a functional change (for better tracking). I would say for these changes you can create two commits, one for fixing the addresses, and one for the new features added.

**Comment by ribdp** [comment]
Changelog - 
1. Created two commits - one for address fixes, and the other for updating the API
2. Used goto error handling instead of returning, in driver init
3. Fixed unused `ret' variable warning, by adding return value checks

**Review by rbolboac — APPROVED** [review]
lgtm

**Review by amiclaus — APPROVED** [review]

**Pull request merged**